### PR TITLE
Add APIs for machine phases

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**: This PR adds the necessary machines-phases and states in the machine-api stack. Machine phases and states are a way of representing the lifecycle of the machines. PR is based on the following proposal doc: [proposal](https://docs.google.com/document/d/12TsBPn1lfMk50_yydzbXNZ9PT-8-88o4tSNi7eqPCVg/edit#)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Added APIs for machine phases
```

### CODE REVIEW

File: pkg/apis/cluster/common/consts.go [hunk 1]
```diff
@@ -103,3 +103,59 @@ const (
...
+// These are the valid values for the MachinePhase.
+const (
+	// MachinePending should be set when machine is being created but it has not joined the
+	// cluster yet.
+	MachinePending MachinePhase = "Pending"
+
+	// MachineRunning should be set when machine has joined the cluster and running successfully.
+	MachineRunning MachinePhase = "Running"
+
+	// MachineRunning should be set when machine is being deleted.
```
Comment (ggaurav10) [inline pkg/apis/cluster/common/consts.go]: Replace MachineRunning with MachineTerminating

Comment (hardikdr) [reply pkg/apis/cluster/common/consts.go]: done, thanks for pointing out.

File: pkg/apis/cluster/common/consts.go [hunk 2]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineOperationHealthCheck should be set when last operation was HealthCheck.
+	MachineOperationHealthCheck MachineOperationType = "HealthCheck"
+
+	// MachineOperationDelete should be set when last operation was Delete.
+	MachineOperationDelete MachineOperationType = "Delete"
+)
+
+// MachineState is the current status of the last performed operation.
+type MachineState string
```
Comment (chaosaffe) [inline pkg/apis/cluster/common/consts.go]: nit: `MachineOperationState` would be a better name as it represents the state of the operation, not of the machine

Comment (hardikdr) [reply pkg/apis/cluster/common/consts.go]: Yes, that makes sense, thanks.
Updating the occurrences wherever needed.

Comment (hardikdr) [reply pkg/apis/cluster/common/consts.go]: done.

**Comment by alvaroaleman** [comment]
Background for the failing tests: The timeout for the test controlplane is too low, so it often fails to start: `2018/10/11 09:27:09 timeout waiting for process kube-apiserver to start`

This already happened multiple times on different PRs. For the time being the workaround is to hit `/retest` until they pass.

I've created a PR in kubebuilder to get this configurable so we can increase the timeout: https://github.com/kubernetes-sigs/controller-runtime/pull/169 It will however take some time until we can use that, because even when that PR is merged we have to wait for the next release.

**Comment by hardikdr** [comment]
thanks, @alvaroaleman for the information, I will then retry after some time.

**Comment by alvaroaleman** [comment]
@hardikdr Actually the failures here a real, check out the logs

**Comment by alvaroaleman** [comment]
@hardikdr Can you quickly explain what this allows us to do what we can currently not do?

Because:

* The machine-controller itself should only take the `.Spec` into account, not the `.Status`
* Other controllers like the `machineSet` controller basically care only about one condition: Is the associated Node existing  and/or ready? They can check that via `.Status.NodeRef` already: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machineset/status.go#L55
* For humans, there are events

Is there anything I missed in the above list? Is there anything I got wrong in the above list?

**Comment by hardikdr** [comment]
Yes sure.
The idea is to basically define the life-cycle of the machines in a descriptive manner.
I see use cases mainly from 2 broad categories: Expressiveness and Consumption.

_1. Expressiveness:_

* Seems it will be nice to have a  framework to express different lifecycle-events of machines as accurately as possible. 

* Taking a subtle example:
Machine gets into the failed-phase. There could be multiple reasons behind it, for instance:
   * Kubelet did not contact APISever from last 10 minutes. 
   * Or Disk-pressure is high from last ~30minutes due to which apps are suffering. 
   * Or simply machine-creation itself has failed due to error either from cloud-provider or say mis-configuration of software-stack on machine.   

 It seems it will be great to express all three scenarios in a structured fashion. 
   * For first 2 cases, we could set MStatus.LastOperation to Health-check and MStatus.LastOparation.State/Description to clearly mention it failed due to kubelet didn't respond in first case and diskPressure was high in second case.

   * The third case is different in a way that creation itself failed, and LastOperation could be then set to `Create` and LO.State/Description could be set : 
      * to saying cloud-provide could not create the machine, 
      * or cloud-provider created the machine but kubelet could not register itself.

* Essentially proposed framework will offer equipped means of expressing the current status-quo of the machine. On top LastOperation field will also provide the latest opreration history of machine-status.

_2. Consumption:_
* Machine could go through many different phases during life-span and machineSetController or User might want to take different actions based on the exact phase.
For instance, 
* MachineSet controller might not want to delete the machine with `Unknown` phase, but rather give some more time to machine to connect back within timeout period.
* MachineSet controller might want to delete the machine immediately to replace it by newer one if its in the Failed phase - assuming machine would had been given enough time already before to join back.
* MachineSet controller might want to ignore machines in Standby phase in case of baremetals, and more.
* Essentially, set of phases should be well-defined for machineSet/other-higher controller to take the right-actions.

To answer the question: `Can you quickly explain what this allows us to do what we can currently not do?`
 * I would say, in current model we do not have means of expressing the phases in detailed and structured way - as described in use case 1 above.
 * And also we do not have well-defined set of inputs for machineSet/other-controllers to take the right actions based on the current status of the machine. - use case 2.

Though machine-shared-controller and machine-external-controller would still require a contract to update the machineStatus, but that's a separate thread. Turned out to be a long one, but I hope it makes sense :) Feedback is most welcome.

**Comment by alvaroaleman** [comment]
Thanks for the detailed answer, @hardikdr !

> It seems it will be great to express all three scenarios in a structured fashion.

We can derive all this information from `.Status.NodeRef`: failed machine creation from the fact that the `.Status.NodeRef` does not exist or does not point to a valid node, the two other examples you mentioned are available as conditions on the `Node` object.

> MachineSet controller might not want to delete the machine with Unknown phase, but rather give some more time to machine to connect back within timeout period.

I don't think it should be the concern of the `MachineSet` controller to re-create cloud provider instances when the node gets lost or never connects, that is something the `Machine` controller does/should do. And again, all this info here is available from the `Machine` object or the referenced `Node` object.

> I would say, in current model we do not have means of expressing the phases in detailed and structured way - as described in use case 1 above

Can you explain why this is needed? For users we can use events, for other controllers do you have an example use-case where the information available via the `.Status.NodeRef` is not sufficient? Because to me it feels like we are duplicating the `Node` object and its features here (conditions, heartbeats)

> And also we do not have well-defined set of inputs for machineSet/other-controllers to take the right actions based on the current status of the machine

Basically the same question as above, do you have a sample of a controller who needs this? The `machineSet` controller only needs to know if a node exists and if its healthy, this works perfectly fine with the existing `.Status.NodeRef`

**Comment by hardikdr** [comment]
thanks for the comments @alvaroaleman .

> We can derive all this information from .Status.NodeRef: failed machine creation from the fact that the .Status.NodeRef does not exist or does not point to a valid node, 

Yes, NodeRef is essential and will provide much info, but I am not sure if there is a way to understand if creation/deletion or other operation has failed or succeeded on a particular machine- via NodeRef's existence. Deriving certain possibilities only based on the availability of NodeRef or pointer to wrong machines- seems less desirable.

>I don't think it should be the concern of the MachineSet controller to re-create cloud provider instances when the node gets lost or never connects, that is something the Machine controller does/should do.

I would rather expect machine-controller should mostly be resposible for creation/deletion of the machines and reporting right health-status via MachineStatus field. MachineSet controller should then look for Failed machines and try to replace them.  This is more from the perspective that MachienSet controller should only make sure to have # of healthy machines and take necessary steps - and not participate in race-condition with MachineController while recreation of machines.

>Can you explain why this is needed? For users we can use events, for other controllers do you have an example use-case where the information available via the .Status.NodeRef is not sufficient? 

Fundamentally, NodeRef is bound by the possible values available in the NodeObject. We would defininitely want to expand our usecases to include new phases such as Draining/Standby[for barementals] and more.

**Comment by davidewatson** [comment]
`MachineOperationType` and `MachineOperationState` can be useful for GUIs built on top of the ClusterAPI. In order to determine if an upgrade is compelte, we currently compare the `Version.Kubelet` and `Version.ControlPlane` with the version reported by the `Node`. This is somewhat limited since kubelet versions are not the only reason for an upgrade.

Others have suggested using Events instead. Can events be lost however, if a controller reboots for example?

From a user perspective I am not sure there is a more reliable way to answer these questions. Is the machine still upgrading, is it being deleted, etc? This is valuable for user feedback within a GUI.

**Comment by roberthbailey** [comment]
As discussed during the meeting today, we will merge this in ~24 hours unless there are objections before then. 

/approve
/hold

File: pkg/apis/cluster/common/consts.go [hunk 3]
```diff
@@ -103,3 +103,59 @@ const (
...
+// These are the valid values for MachineOperationState.
+const (
+	// MachineOperationStateProcessing should be set when last operation is ongoing.
+	MachineOperationStateProcessing MachineOperationState = "Processing"
+
+	// MachineOperationStateFailed should be set when last operation has failed with certain error.
+	MachineOperationStateFailed MachineOperationState = "Failed"
+
+	// MachineOperationStateSuccessful should be set when last operation was successfully completed.
+	MachineOperationStateSuccessful MachineOperationState = "Successful"
```
[the remark this answers is not in the document]
Comment (detiber) [reply pkg/apis/cluster/common/consts.go]: `Succeed` seems a bit odd in context here. Any reason to not use `Success`?

[the remark this answers is not in the document]
Comment (chaosaffe) [reply pkg/apis/cluster/common/consts.go]: It felt weird to me too, but I couldn't decide if the pair should be 
`Success` and `Failure`
or 
`Succeeded` and `Failed`

[the remark this answers is not in the document]
Comment (detiber) [reply pkg/apis/cluster/common/consts.go]: Personally, I think I prefer `Success` and `Failure`, but s/Succeed/Succeeded would be a bit more consistent as well.

File: pkg/apis/cluster/common/consts.go [hunk 4]
```diff
@@ -103,3 +103,59 @@ const (
...
+type MachinePhase string
+
+// These are the valid values for the MachinePhase.
+const (
+	// MachinePending should be set when machine is being created but it has not joined the
+	// cluster yet.
+	MachinePending MachinePhase = "Pending"
+
+	// MachineRunning should be set when machine has joined the cluster and running successfully.
+	MachineRunning MachinePhase = "Running"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: This is the same as `Machine` exists and has a `.Status.NodeRef` that points to a healthy node

File: pkg/apis/cluster/common/consts.go [hunk 5]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineTerminating should be set when machine is being deleted.
+	MachineTerminating MachinePhase = "Terminating"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: This is the same as `MachineExists` and has a `DeletionTimestamp` that is not nil

File: pkg/apis/cluster/common/consts.go [hunk 6]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineUnknown should be set when there is no way for controller to understand the
+	// current phase of the machine. For example when is not reachable due to network-issues.
+	MachineUnknown MachinePhase = "Unknown"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: This is the same as `Machine` exists and has a `.Status.NodeRef` that points to an unhealthy node

File: pkg/apis/cluster/common/consts.go [hunk 7]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineFailed should be set when machine has failed and certain action needs to be taken either by higher-level controller or an admin.
+	MachineFailed MachinePhase = "Failed"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: Why not simply use `.Status.ErrorReason` and `.Status.ErrorMessage` for this when targeting humans?

Other controllers like the `MachineSet` controller already use [`.Status.NodeRef`]( https://github.com/kubernetes-sigs/cluster-api/blob/fee897706a82938117591515fbd4e85e667299a7/pkg/controller/machineset/status.go#L55-L60) and IMHO should continue to do so as [node health checking is a tricky problem we should not re-implement](https://github.com/kubernetes/community/blob/master/keps/sig-node/0009-node-heartbeat.md)

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 8]
```diff
@@ -144,6 +144,33 @@ type MachineStatus struct {
...
+
+	// LastOperation describes the last-operation performed by the machine-controller.
+	// This API should be useful as a history in terms of the latest operation performed on the
+	// specific machine. It should also convey the state of the latest-operation for example if // it is still on-going, failed or completed successfully.
+	// +optional
+	LastOperation LastOperation `json:"lastOperation,omitempty"`
+
+	// Phase represents an ongoing-phase a machine is going through. It is expected to be
+	// transient. Higher-level controllers should rely on MachinePhases to take the right
+	// decision such as replacing the Failed machines.
```
Comment (alvaroaleman) [inline pkg/apis/cluster/v1alpha1/machine_types.go]: I disagree. Higher-Level controllers should rely on `.Status.NodeRef` as [node health checking is a tricky problem](https://github.com/kubernetes/community/blob/master/keps/sig-node/0009-node-heartbeat.md)

File: pkg/apis/cluster/common/consts.go [hunk 9]
```diff
@@ -103,3 +103,59 @@ const (
...
+)
+
+// MachineOperationType is the type of the last operation performed on the Machine.
+type MachineOperationType string
+
+// These are valid values for MachineOperationType.
+const (
+	// MachineOperationCreate should be set when last operation was Create.
+	MachineOperationCreate MachineOperationType = "Create"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: The same as `Machine` exists and has no `.Status.NodeRef`

File: pkg/apis/cluster/common/consts.go [hunk 10]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineOperationUpdate should be set when last operation was Update.
+	MachineOperationUpdate MachineOperationType = "Update"
+
+	// MachineOperationHealthCheck should be set when last operation was HealthCheck.
+	MachineOperationHealthCheck MachineOperationType = "HealthCheck"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: The same as `Machine` exists and has a `.Status.NodeRef` pointing to a healthy node

File: pkg/apis/cluster/common/consts.go [hunk 11]
```diff
@@ -103,3 +103,59 @@ const (
...
+
+	// MachineOperationDelete should be set when last operation was Delete.
+	MachineOperationDelete MachineOperationType = "Delete"
```
Comment (alvaroaleman) [inline pkg/apis/cluster/common/consts.go]: The same as `Machine` exists and has a non-nil `DeletionTimestamp`

**Comment by chaosaffe** [comment]
@alvaroaleman thanks for the detailed breakdown of how we can get all this information with the current API!

File: pkg/apis/cluster/common/consts.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chaosaffe) [reply pkg/apis/cluster/common/consts.go]: Sorry, that was just a brain stutter on my part. Thanks for catching it @detiber

**Comment by hardikdr** [comment]
Basically, it is boiling down to 2 parallel approaches:
1. Using NodeRef to infer the machine-health/machine-status. 
   * This is also the approach we are currently using in the implementation.
2. Using MachinePhases/State proposed in the PR.
   * Add new fields to convey the current machine-phase and last-operation details.

Though I have typed the phases/states in PR, It seems we could remove them and keep the format only  to "string" for machine-phase and states [[ref](https://github.com/kubernetes-sigs/cluster-api/blob/d58060cf29c96b9cf5f0653aaf29bd5e22a20031/pkg/apis/cluster/common/consts.go#L111-L161)]. The question would still remain if we need the framework or not.
We will internally be brainstorming on this by end of the week, would be nice if more folks here could provide their views.

File: pkg/apis/cluster/common/consts.go (hunk 10 above)
Comment (enxebre) [inline pkg/apis/cluster/common/consts.go]: what would be setting a MachineOperationHealthCheck MachineOperationType? is this the machine controller monitoring the node status periodically? is this a gate for custom controllers?

**Comment by sidharthsurana** [comment]
Another use case to keep in mind is when a **_central_** cluster is managing number of **_remote_** clusters. The `NodeRef` as of today only is applicable for a self-managed cluster and not present otherwise. So unless we extend the current `NodeRef` bookkeeping behavior to **remotely managed** cluster as well, all the machine lifecycle states/phases mentioned above cannot be derived for remote clusters. This is one case when I think the states/phases in this proposal could help unless we choose to extend the `NodeRef` behavior as above.

**Comment by alvaroaleman** [comment]
> Another use case to keep in mind is when a central cluster is managing number of remote clusters. The NodeRef as of today only is applicable for a self-managed cluster and not present otherwise. So unless we extend the current NodeRef bookkeeping behavior to remotely managed cluster as well, all the machine lifecycle states/phases mentioned above cannot be derived for remote clusters. This is one case when I think the states/phases in this proposal could help unless we choose to extend the NodeRef behavior as above.

That is indeed a very valid use case, however I believe the phases are not the solution for that as that would basically mean we re-implement the Node heartbeating. There is btw. already an issue for remote clusters: #520

**Comment by davidewatson** [comment]
I agree that the lack of `NodeRef`s in the remote cluster case is a larger problem and need not be addressed within this PR. I still think we should have explicit fields for consumers. This does require a process to update the fields, but that is better than asking a person to do that work.

I would be okay with just using strings, though I don't think this really buys us any freedom to change the API. It's probably better for consumers if the reported values are consistent.

**Comment by hardikdr** [comment]
@chaosaffe  Today we briefly discussed this PR again, and I have made a change of removing the typed-strings from API. Can you please have a quick look.
cc: @alvaroaleman

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 12]
```diff
@@ -151,6 +151,36 @@ type MachineStatus struct {
 	// Refer: https://kubernetes.io/docs/concepts/architecture/nodes/#condition
 	// +optional
 	Conditions []corev1.NodeCondition `json:"conditions,omitempty"`
+
+	// LastOperation describes the last-operation performed by the machine-controller.
+	// This API should be useful as a history in terms of the latest operation performed on the
+	// specific machine. It should also convey the state of the latest-operation for example if // it is still on-going, failed or completed successfully.
```
Comment (chaosaffe) [inline pkg/apis/cluster/v1alpha1/machine_types.go]:
```suggestion
	// specific machine. It should also convey the state of the latest-operation for example if 
	// it is still on-going, failed or completed successfully.
```

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 13]
```diff
@@ -151,6 +151,36 @@ type MachineStatus struct {
...
+	// +optional
+	LastOperation LastOperation `json:"lastOperation,omitempty"`
+
+	// Phase represents an ongoing-phase a machine is going through. It is expected to be
+	// transient. Higher-level controllers should rely on MachinePhases to take the right
+	// decision such as replacing the Failed machines.
```
Comment (chaosaffe) [inline pkg/apis/cluster/v1alpha1/machine_types.go]:
```suggestion
```

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 14]
```diff
@@ -151,6 +151,36 @@ type MachineStatus struct {
...
 	Conditions []corev1.NodeCondition `json:"conditions,omitempty"`
+
+	// LastOperation describes the last-operation performed by the machine-controller.
+	// This API should be useful as a history in terms of the latest operation performed on the
+	// specific machine. It should also convey the state of the latest-operation for example if // it is still on-going, failed or completed successfully.
+	// +optional
+	LastOperation LastOperation `json:"lastOperation,omitempty"`
+
+	// Phase represents an ongoing-phase a machine is going through. It is expected to be
+	// transient. Higher-level controllers should rely on MachinePhases to take the right
```
Comment (chaosaffe) [inline pkg/apis/cluster/v1alpha1/machine_types.go]:
```suggestion
```

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 15]
```diff
@@ -151,6 +151,36 @@ type MachineStatus struct {
...
 	// +optional
 	Conditions []corev1.NodeCondition `json:"conditions,omitempty"`
+
+	// LastOperation describes the last-operation performed by the machine-controller.
+	// This API should be useful as a history in terms of the latest operation performed on the
+	// specific machine. It should also convey the state of the latest-operation for example if // it is still on-going, failed or completed successfully.
+	// +optional
+	LastOperation LastOperation `json:"lastOperation,omitempty"`
+
+	// Phase represents an ongoing-phase a machine is going through. It is expected to be
```
Comment (chaosaffe) [inline pkg/apis/cluster/v1alpha1/machine_types.go]:
```suggestion
	// Phase represents the current phase of machine actuation.
```

**Comment by hardikdr** [comment]
@chaosaffe thanks for the suggestions. All of them looks good to me. I also made the necessary changes.

File: pkg/apis/cluster/v1alpha1/machine_types.go [hunk 16]
```diff
@@ -151,6 +151,35 @@ type MachineStatus struct {
 	// Refer: https://kubernetes.io/docs/concepts/architecture/nodes/#condition
 	// +optional
 	Conditions []corev1.NodeCondition `json:"conditions,omitempty"`
+
+	// LastOperation describes the last-operation performed by the machine-controller.
+	// This API should be useful as a history in terms of the latest operation performed on the
+	// specific machine. It should also convey the state of the latest-operation for example if 
+	// it is still on-going, failed or completed successfully.
+	// +optional
+	LastOperation LastOperation `json:"lastOperation,omitempty"`
```
Comment (alvaroaleman) [inline pkg/apis/cluster/v1alpha1/machine_types.go]: Please make this a pointer as its an optional struct

Comment (hardikdr) [reply pkg/apis/cluster/v1alpha1/machine_types.go]: yes sure, done.

**Comment by roberthbailey** [comment]
It looks like we've pretty much reached consensus on this PR. Let's make sure there are no further comments or objections at the next meeting and then merge it if all looks good.

File: config/crds/cluster_v1alpha1_machine.yaml [hunk 17]
```diff
@@ -71,11 +71,25 @@ spec:
               type: string
             errorReason:
               type: string
+            lastOperation:
+              properties:
+                description:
+                  type: string
+                lastUpdateTime:
```
[the remark this answers is not in the document]
Comment (hardikdr) [reply config/crds/cluster_v1alpha1_machine.yaml]: done, thanks.

**Comment by derekwaynecarr** [comment]
I apologize for the late question, but is there a write-up that summarizes a response from Daniel's feedback here:
https://docs.google.com/document/d/12TsBPn1lfMk50_yydzbXNZ9PT-8-88o4tSNi7eqPCVg/edit?disco=AAAACWAG9QI 

similar to him, I worry about the use of a single phase based on experience with pods.

**Comment by roberthbailey** [comment]
@hardikdr -- can you answer @derekwaynecarr's question? I know that you had a chance to connect with @lavalamp at KubeCon and chat about phases / conditions / etc.
